### DISCUSSION TITLE
Add/append/replace binary data

### DISCUSSION THREAD

**Discussion by floitsch** [lief-project/LIEF#756] [Q&A] [answered]
Context:
We have a virtual machine that interprets bytecodes.
Our users use the compiler to compile their programs, which can then be executed by the interpreter.
I'm now trying to allow the user to bundle the interpreter and the bytecodes into one self-contained executable.
We are running on Linux, Windows, macOS (but would like to target other platforms in the future).

My hope is that I can link LIEF into the compiler and then update a copy of the interpreter to add/append/replace the bytecodes there.
Can this be done with LIEF, and if yes, could someone point me to the functions I should use? (c++)

Initially I was planning to just append the bytecodes to the interpreter without even modifying the executable. I could then use a library like [whereami](https://github.com/gpakosz/whereami) to find my own executable and extract the bytecodes from there.
However this approach doesn't work nicely with signing as the bytecodes wouldn't be part of the executable. If I could add the bytecodes with LIEF that would be much cleaner.

I see two possibilities (depending on your recommendations):
- use LIEF to add the bytecodes at the end of the executable, and use `whereami` to extract them. In this case, LIEF would just make sure that the data is at the end of the file and is still part of the executable. The interpreter would not have a symbol for it, but read itself and look for magic numbers at the end of the executable.
- generate the interpreter with a dummy bytecode-data, and then replace the data with LIEF. In this case, I would look up the symbol and replace the data. I'm not sure if LIEF supports this, though.

**Comment by romainthomas** [comment, marked as answer]
Hello @floitsch,
Firstly, I would say that the LIEF's PE support for modifying binaries is not really reliable compared to ELF/Mach-O. I plan to refactor this part by the end of the year but keep mind this aspect if you want to use it in a product. 

From what I understood of your problem, I think your second approach (creating an empty space for the bytecode) would be 
more reliable. Actually you can do that with this kind of code:

```python
import lief

target = lief.parse("...")
sym = target.get_symbol("__bytecode_data")
target.patch_address(sym.value, [your new bytecode])
target.write("out.bin")
```

**Reply by floitsch** [reply]
Thanks!
I will give it a try and report back.

**Reply by floitsch** [reply]
Played with it today.
Patching the address worked great. However, it requires the space to be available.

I played a bit with resizing the executable, but that turned out to be more complicated than I thought. I think I will start by just patching existing space.
